r/src/dataset.cpp

```diff
@@ -76,8 +76,14 @@ std::vector<std::string> dataset___FileSystemDataset__files(
 
 // [[arrow::export]]
 std::shared_ptr<ds::Dataset> dataset___DatasetFactory__Finish1(
-    const std::shared_ptr<ds::DatasetFactory>& factory) {
-  return VALUE_OR_STOP(factory->Finish());
+    const std::shared_ptr<ds::DatasetFactory>& factory, bool unify_schemas) {
+  ds::FinishOptions opts;
+  if (unify_schemas) {
+    ds::InspectOptions in_opts;
+    opts.inspect_options.fragments = in_opts.kInspectAllFragments;
+    opts.validate_fragments = true;
```

Comment:
Validating all fragments against a schema should only be necessary if the dataset schema *hadn't* been inspected and unified from all fragments. I'd expect this to either be an independent argument to `dataset___DatasetFactory__Finish1` or dropped
```suggestion
```

Reply:
But we don’t do Inspect first I don’t think, we just call Finish

Reply:
If schema is not explicitly set in FinishOptions, then `inspect_options` is used to invoke Inspect automatically in order to determine a schema. This happens in C++:
https://github.com/apache/arrow/blob/303bcc2b887f1b8b97993f15941370aedb2e67e1/cpp/src/arrow/dataset/discovery.cc#L92-L95

Since you just inspected and unified the schemas of all the fragments anyway, it's redundant to then verify them against that unified schema

Reply:
In order to validate schemas, you'd probably want to add `validate_schemas` as a parameter of `dataset___DatasetFactory__Finish2` (if TRUE, check that all fragments match the schema which was explicitly provided)